### PR TITLE
KFX support

### PR DESCRIPTION
**Pull request by tomthumb1997**
As requested by apprenticeharper here are changes to integrate KFX support.

ion.py does the bulk of the work. (Created by lulzkabulz and converted to Python by apprenticenaomi.)
kfxdedrm.py handles the set of files that make up a book in KFX format as a ZIP archive..
Misc changes made to __init__.py, k4mobidedrm.py, and kgenpids.py to make it work.

Tested as best I can. Hope this helps.

### CODE REVIEW

**Comment by voltagex** [comment]
I've tested this by packing up the files into a plugin and loading it into Calibre - the log looks good but the book is still in kfx-zip format in my library - is the format meant to change too?

**Comment by j-howell** [comment]
> I've tested this by packing up the files into a plugin and loading it into Calibre

I haven't had a chance to test it yet. Could you post the plugin zip that includes the patches somewhere? That would save the rest of us some time.

>  the log looks good but the book is still in kfx-zip format in my library 

That log looks like something happened. You could try converting the result with the [KFX Input plugin](https://www.mobileread.com/forums/showthread.php?t=291290) for calibre. That will only succeed if DRM has been removed.

> is the format meant to change too?

I put a feature in the KFX Input plugin that is intended to change the format from KFX-ZIP to (monolithic) KFX if the import of a KFX book without DRM succeeds, but I was not able to test it properly due to lack of DRM removal in the Tools. Perhaps I made a mistake in that code.

Do you have that plugin installed?

**Comment by voltagex** [comment]
@j-howell Thanks for your response. I can confirm that with the KFX Input plugin it works as expected - book and metadata are imported and I'm able to convert from KFX to Mobi without an issue.

I think this is showing a bit of a gap in Calibre's capabilities - plugins need to be able to send errors to the UI so I'd know (for example) to go and download the plugin.

Anyway, sorry for the extra noise on your PR - perhaps the easiest solution is for the KFX Plugin to be bundled with DeDRM.

**Comment by j-howell** [comment]
> I can confirm that with the KFX Input plugin installed it works as expected - book and metadata are imported and I'm able to convert from KFX to Mobi without an issue.

Great. Thanks for trying it and for the zip file. I want to give it a try myself and see if I can find any problems.

> I think this is showing a bit of a gap in Calibre's capabilities - plugins need to be able to send errors to the UI so I'd know (for example) to go and download the plugin.

I agree! I have often thought it would be nice for DRM removal failures to be reported to the user in a popup dialog box when they occur instead of the user needing to recognize that an imported book still has DRM and then have to generate a log in debug mode.

I wonder if there is something in the architecture of calibre that prevents this or if it is a matter of no one taking the time to implement that feature.

> perhaps the easiest solution is for the KFX Plugin to be bundled with DeDRM.

I don't see a need for that and I want to keep my efforts in converting KFX separate from these tools since DRM removal is a gray area legally in many places.

File: DeDRM_calibre_plugin/DeDRM_plugin/k4mobidedrm.py [hunk 1]
```diff
@@ -197,13 +200,15 @@ def GetDecryptedBook(infile, kDatabases, androidFiles, serials, pids, starttime
     mobi = True
     magic8 = open(infile,'rb').read(8)
     if magic8 == '\xeaDRMION\xee':
-        raise DrmException(u"KFX format detected. This format cannot be decrypted yet.")
-        
+        raise DrmException(u"The .kfx DRMION file cannot be decrypted by itself. A .kfx-zip archive containing a DRM voucher is required.")
```
[the remark this answers is not in the document]
Comment (mafrosis) [reply DeDRM_calibre_plugin/DeDRM_plugin/k4mobidedrm.py]: I'm currently working on a patch for that :)

[the remark this answers is not in the document]
Comment (cemeyer) [reply DeDRM_calibre_plugin/DeDRM_plugin/k4mobidedrm.py]: Great! :-)

**Pull request merged**